shell/common/shell.cc

```diff
@@ -1762,12 +1762,23 @@ fml::Status Shell::WaitForFirstFrame(fml::TimeDelta timeout) {
                        "because it is responsible for generating the frame.");
   }
 
+  // Check for overflow.
```

Comment:
@gaaclarke - mind taking another look at this?

I had to change this because otherwise we're susceptible to overflows when adding `timeout` to `std::chrono::steady_clock::now()`.

`wait_for` does that addition behind the scenes and I guess treats the overflow as UB. I'm using wait_until to avoid the steady clock advancing while we do other calcuations and overflowing even though we checked.

This is covered by tests now that the timeouts use max/zero values.

Reply:
FWIW, this is an actual bug for us today - if someone were to pass a very large value into here, they'd get a timeout response due to overflow.

Reply:
It sounds like the API docs in the header file here https://github.com/flutter/engine/blob/master/shell/common/shell.h#L301 could use some more details or a warning about this.

Reply:
Adding a doc.

Reply:
Good catch.  I think an easier fix would just be to say `auto duration = min(max_duration, desired_duration);` and use wait_for.  Since that time is so long, no one will ever be able to tell the difference in their lifetime.  That removes a branch and is much easier to verify it is correct assuming what was there before was correct.

Reply:
Done. Though technically isn't std::min a branch? :) Anyway I agree the code is more readable/maintainable this way.

Reply:
Oh. We can't do that. `wait_for` is still going to add to `stead_clock::now()`, which has incremented since checked for overflow, and now might overflow again.

Reply:
We could do this:

```cpp
  auto now = std::chrono::steady_clock::now();
  auto max_duration = std::chrono::steady_clock::time_point::max() - now;
  auto desired_duration = std::chrono::milliseconds(timeout.ToMilliseconds());

  std::unique_lock<std::mutex> lock(waiting_for_first_frame_mutex_);
  bool success = waiting_for_first_frame_condition_.wait_until(
      lock, now + std::min(desired_duration, max_duration),
      [&waiting_for_first_frame = waiting_for_first_frame_] {
        return !waiting_for_first_frame.load();
      });
```

Reply:
That sounds good to me too, as long as there isn't that branch I think it's more straightforward.